File: framework/test2/validate/list2d/list2d.py
#!/usr/bin/env python3
from collections import Counter, defaultdict, deque
import heapq
import json
import logging
import math
import random
import time
from typing import *
import sys

from soda.leetcode.bitree import *
from soda.leetcode.graph import *
from soda.leetcode.linklist import *
from soda.leetcode.nest import *
from soda.unittest.util import init_logging

logger = logging.getLogger(__name__)

# step [1]: implement class Solution
# class Solution: pass
class Solution:
    def groupByLength(self, strs: List[str]) -> List[List[str]]:
        random.shuffle(strs)
        group = defaultdict(list)
        for s in strs:
            group[len(s)].append(s)
        keys = list(group.keys())
        random.shuffle(keys)
        return list(map(lambda x: group[x], keys))

if __name__ == '__main__':
    init_logging()
    from soda.unittest import Validators
    from soda.unittest.work import TestWork

    # step [2]: setup function
    # Attention! FUNCTION must use type hint, including arguments and return type
    work = TestWork(Solution().groupByLength)
    # OR use struct tester
    # work = TestWork.forStruct(CLASS)

    # step [3]: setup other options
    work.validator = Validators.forList2d(str, False, False)
    work.compareSerial = True
    work.run()

File: framework/test2/validate/list2d/list2d.php
<?php
require_once 'Soda/DS/index.php';
require_once 'Soda/Leetcode/index.php';
require_once 'Soda/Unittest/index.php';

use Soda\DS;
use Soda\Leetcode\ListNode;
use Soda\Leetcode\NestedInteger;
use Soda\Leetcode\TreeNode;
use Soda\Unittest\TestWork;
use Soda\Unittest\Utils;
use Soda\Unittest\Validators;

class Solution {

    /**
     * @param String[] $strs
     * @return String[][]
     */
    function groupByLength($strs) {
        shuffle($strs);
        $group = array();
        foreach ($strs as $s) {
            $group[strlen($s)][] = $s;
        }
        $keys = array_keys($group);
        shuffle($keys);
        return array_map(fn($k) => $group[$k], $keys);
    }
}

$work = TestWork::create(new Solution(), 'groupByLength', $argv[0]);
// $work = TestWork::forStruct('STRUCT::class', $argv[0]);
$work->validator = Validators::forArray2d('String', false, false);
$work->compareSerial = true;
echo $work->run(Utils::fromStdin());

